docs/proposals/resource-quota-reservations.md

```diff
@@ -0,0 +1,228 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Resource Quota - reservations
+
+## Problem Description
+
+When a request for creating a resource is received,
+the associated resource quota is checked by the
+server in admission control.  If the request would
+not violate the quota, the quota usage is incremented
+and the requesting object is allowed to be created.
+Quota usage is incremented using compare-and-swap
+to allow for optimistic locking.  If the quota usage
+was stale, the admission control logic runs again
+to validate the request is still valid with the latest
+quota document.
+
+A quota controller regularly sweeps the system to
+recalculate observed usage.  This is primarily useful
+for monitoring deleted resources and replenishing
+available quota.  If the quota controller sees a delta
+in its recalculated usage relative to the currently
+reported status, it will update the usage.  If its update
+is stale, it will recalculate usage again.
+
+If the quota controller works has an up to date quota,
+but interacts with a latent API server, the quota controller
+could under report usage.  This is possible in a single
+server environment (though the race window is small), but
+the risk is commensurate to the number of masters in an
+HA configuration and their relative latency.
+
+In pseudo-code:
+
+**API_SERVER**
+
+```
+Admission control (quota)
+T1. QUOTA = ...
+T2. IF QUOTA.USED + REQUEST >= QUOTA.HARD
+T3.   REJECT
+T4. UPDATE QUOTA.STATUS(QUOTA.USED + REQUEST)
+T5. ADMIT
+
+RESTStorage
+T6. Create Object
+```
+
+**QUOTA_CONTROLLER**
+
+```
+T1. QUOTA := ...
+T2. OBSERVED_USAGE := ... [live queries based on latest data]
+T3. UPDATE QUOTA.STATUS.USED IFF OBSERVED USAGE != QUOTA.STATUS.USED
+```
+
+If the `QUOTA_CONTROLLER.T1` happens after `API_SERVER.T4`,
+and `QUOTA_CONTROLLER.T3` happens before `API_SERVER.T6` completes,
+then when the quota controller recalculates usage, it could have calculated
+usage without observing the outcome of the create.
```

Comment:
There must be something about ResourceVersion I don't understand. If the quota controller writes quota.status in between API_SERVER.T4 and API_SERVER.T6, won't API_SERVER.T6 fail due to resource version conflict? These are all writing the same object, right?